### PR TITLE
Bugfix for #14: Fix interface configuration for WildFly 9 and EAP 7

### PR DESCRIPTION
**Pull request by therek**
Avoid `any-ipv4-address` and `any-ipv6-address` parameters for Wildfly version 9.x and newer. Bugfix for #14.

### CODE REVIEW

**Comment by therek** [comment]
Please review @coi-gov-pl/jboss-puppet-team

**Comment by cardil** [comment]
I'm not sure but I think that JBoss EAP starting from version 6.4 should also be included in your changes. 

We need to research that.

**Comment by cardil** [comment]
This needs double check but it looks like JBoss EAP 6.4.x has everything as it was. At least according to documentation: https://access.redhat.com/documentation/en-US/JBoss_Enterprise_Application_Platform/6.4/html/Administration_and_Configuration_Guide/chap-Network_and_Port_Configuration.html#sect-Interfaces

JBoss EAP 7 Beta has differently written docs with chapter for IPv6 addresses: https://access.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/7.0.beta/configuration-guide/chapter-4-network-and-port-configuration

Please double check that @therek on real installation, ok?

File: manifests/interface.pp [hunk 1]
```diff
@@ -134,6 +132,18 @@
     'virtual'            => $virtual,           # undef, bool
   }
 
+  $legacy_bind_variables = {
+    'any-ipv4-address'   => $any_ipv4_address,  # undef, bool
+    'any-ipv6-address'   => $any_ipv6_address,  # undef, bool
+  }
+
+  if $jboss::product == 'wildfly' and versioncmp($jboss::version, '9.0.0') >= 0 {
+    $bind_variables = $basic_bind_variables
```
Comment (cardil) [inline manifests/interface.pp]: We should throw here a deprecated warning if values are set to something other than undef and using newer versions of products

File: spec/defines/interface_spec.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  context 'with jboss_running => true and runasdomain => false parameters set' do
+    let(:runtime_facts) do
+      { :jboss_running => 'true' }
+    end
+    let(:params) do
+      { :runasdomain => 'false' }
+    end
+    let(:pre_condition) { "class { jboss: product => '#{product}', version => '#{version}'}" }
+
+    context 'with product => wildfly and version => 9.0.2.Final parameters set' do
```
Comment (cardil) [inline spec/defines/interface_spec.rb]: Test also for not compatible versions. For ex wildfly 15 should `raise_error (Puppet::Error, /Unsupported version Wildfly 15/)`

**Comment by therek** [comment]
Verified configuration options:
- JBoss EAP 6.4.0.GA: supports `any-ipv4-address` and `any-ipv6-address`;
- JBoss EAP 7.0.0.Beta: doesn't support `any-ipv4-address` or `any-ipv6-address`.

**Comment by cardil** [comment]
For me it's cool! Gr8 PR!

**Comment by therek** [comment]
Waiting to be merged and released.

**Comment by cardil** [comment]
I think we should set WildFly 9.0.2.Final as default. Now it is outdated 8.2.0.Final. Also it will be great to test this with upcoming WildFly 10.0.0.CR5.

Only question is: do it in this PR or seperate? @therek what do you think? I'm opting for doing it here.

**Comment by therek** [comment]
I've bumped default version to 9.0.2.Final. Tests for Wildfly 10 can be done separately.

**Comment by cardil** [comment]
I rebased this PR' branch to current develop

**Comment by cardil** [comment]
It's seams to be some kind of error in acceptance test running on beaker'ed Docker container for Debian 7. I think this is because there is old version of Augeas in Debian 7 and it's not upgraded in nodeset: https://github.com/coi-gov-pl/puppet-jboss/blob/7234c7c5f1945f6e6264a7709c115b130ca99084/spec/acceptance/nodesets/debian-7-x86_64-docker.yml

I don't know how this could work in other PR's :confused:

File: manifests/interface.pp (hunk 1 above)
Comment (cardil) [inline manifests/interface.pp]: !fixed

File: spec/defines/interface_spec.rb (hunk 2 above)
Comment (cardil) [inline spec/defines/interface_spec.rb]: !fixed

**Comment by cardil** [comment]
Problem persists on Debian 7 docker instance. It's  repeating so it is not a network fluke. @therek maybe run a Debian 7 docker image for your self and try to reproduce the error? More detective work is needed.

**Comment by cardil** [comment]
I've executed it on Debian:7 container with `--debug` flag. It clears to me that there is a Augeas error:

```
Debug: Augeas[ensure present interface public](provider=augeas): /augeas/files/usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml/error = parse_failed
Debug: Augeas[ensure present interface public](provider=augeas): /augeas/files/usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml/error/pos = 1
Debug: Augeas[ensure present interface public](provider=augeas): /augeas/files/usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml/error/line = 1
Debug: Augeas[ensure present interface public](provider=augeas): /augeas/files/usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml/error/char = 1
Debug: Augeas[ensure present interface public](provider=augeas): /augeas/files/usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml/error/lens = /usr/lib/wildfly-9.0.2.Final/lenses/jbxml.aug:136.10-138.60:
```

This is it in lens file (it's our file, supplied by @Garagoth). Error is reported to be here: https://github.com/coi-gov-pl/puppet-jboss/blob/develop/files/jbxml.aug#L136-L138 but I think that this file is copied from newer version of augeas by @Garagoth without any changes. Usually this is caused by changes to XML document but it's also strange because it seams to work with wildfly-9.0.2 on Ubuntu and Centos.

Maybe package versions of augeas will be helpful:

```
# dpkg -l | grep augeas
ii  augeas-lenses                   0.10.0-1                   all          Set of lenses needed by libaugeas0 to parse config files
ii  libaugeas-ruby                  0.4.1-1.1                  all          Augeas bindings for the Ruby language
ii  libaugeas-ruby1.9.1             0.4.1-1.1                  amd64        Augeas bindings for the Ruby language
ii  libaugeas0                      0.10.0-1                   amd64        Augeas configuration editing library and API
```

Here is a head of failing XML file. I'm pretty sure that suspicious looking header `<?xml version="1.0" ?>` with conjunction to specific version of augeas is causing this. It's technically valid but more popular should be `<?xml version="1.0" encoding="UTF-8" standalone="no" ?>`

```
# head /usr/lib/wildfly-9.0.2.Final/standalone/configuration/standalone-full.xml
<?xml version="1.0" ?>

<server xmlns="urn:jboss:domain:3.0">
    <extensions>
        <extension module="org.jboss.as.clustering.infinispan"/>
        <extension module="org.jboss.as.connector"/>
        <extension module="org.jboss.as.deployment-scanner"/>
        <extension module="org.jboss.as.ee"/>
        <extension module="org.jboss.as.ejb3"/>
        <extension module="org.jboss.as.jaxrs"/>
```
**UPDATE:** I've checked my XML header theory by changing `<?xml version="1.0" ?>` to other forms but it doesn't seems to fix the problem. More sniffing is needed :cry:

**Comment by cardil** [comment]
I've merged newest develop. Now still debian:7 build fails: https://travis-ci.org/coi-gov-pl/puppet-jboss/builds/107445772

**Comment by therek** [comment]
To summarize the Debian 7 issue: the problem was with Jboss configuration file's header that in Wildfly 8.x looks like this
`<?xml version='1.0' encoding='UTF-8'?>`
while in Wildfly 9.x
`<?xml version="1.0" ?>`

The difference is, that 9.x uses double-quotes instead of single-quotes. Augeas lens used for parsing this file assumed there's a single-quote here and failed on Augeas version 0.10.0 available on Debian 7. Dunno why it worked for newer versions on Ubuntu/Centos.

The actual change to `(sto_dquote|sto_squote)` says that the attributes in XML header can be eighter double- or single-quoted, respectively (`sto_dquote` and `sto_squote` are defined near the beginning of the lens file).

**Comment by cardil** [comment]
Good description. :+1: Maybe we should create an issue to determine why only Debian being affected. I don't like leaving that in the darkness. What do you think @therek?

**Comment by therek** [comment]
That's actually easy to answer and you, of all, should know, as this comes from the code you wrote, my dear Watson :)

Ubuntu repository provides Augeas version 1.2.0, while CentOS's repo version 1.0.0. Both this versions contain XML lens that[ can parse single- and double-quotes](https://github.com/hercules-team/augeas/commit/ea3208762e0b9e0b55d830916b36cda3c72c06c7) in the header, though a bit differently from what I proposed (Augeas v0.10.0 doesn't support the syntax used in later version). Now, in the [augeas.pp](puppet-jboss/manifests/internal/augeas.pp) manifest there's an if statement that uses jbxml.aug for older versions of Augeas and native XML lens file for versions >= 1.0.0.

Mystery solved, my dear Watson. We can rest now :)

**Comment by cardil** [comment]
Oh. That's it. Now I think you have also fixed bug for Puppet Enterprise < 3.0. Augers there is bundled with PE and is old.

Double nice!

**Pull request merged**